File: node-services/auth-service/idempotency.js
import { getPool } from './db.js';

export function idempotencyMiddleware() {
  return async (req, res, next) => {
    const key = req.headers['idempotency-key'];
    if (!key) return next();

    const pool = getPool();
    if (!pool) return next();

    try {
      const result = await pool.query(
        'SELECT response, status_code FROM idempotency_keys WHERE key = $1 AND expires_at > NOW()',
        [key],
      );
      if (result.rows.length > 0) {
        const cached = result.rows[0];
        return res.status(cached.status_code).json(JSON.parse(cached.response));
      }
    } catch {
      // store unavailable, proceed without idempotency
    }

    const originalJson = res.json.bind(res);
    res.json = (body) => {
      pool
        ?.query(
          `INSERT INTO idempotency_keys (key, response, status_code, expires_at)
         VALUES ($1, $2, $3, NOW() + INTERVAL '24 hours')
         ON CONFLICT (key) DO UPDATE SET response = $2, status_code = $3`,
          [key, JSON.stringify(body), res.statusCode],
        )
        .catch(() => {});
      return originalJson(body);
    };
    next();
  };
}

File: services/internal/platform/idempotency.go
package platform

import (
	"context"
	"errors"
	"log/slog"
	"time"

	"connectrpc.com/connect"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

func ExtractIdempotencyKey(headers map[string]string) string {
	return headers["Idempotency-Key"]
}

type IdempotencyStore struct {
	pool *pgxpool.Pool
}

func NewIdempotencyStore(pool *pgxpool.Pool) *IdempotencyStore {
	if pool == nil {
		return nil
	}
	return &IdempotencyStore{pool: pool}
}

// Check returns cached response if key exists and not expired. Returns nil if not found.
func (s *IdempotencyStore) Check(ctx context.Context, key string) ([]byte, int, bool, error) {
	if s == nil || key == "" {
		return nil, 0, false, nil
	}
	var response []byte
	var statusCode int
	err := s.pool.QueryRow(ctx,
		"SELECT response, status_code FROM idempotency_keys WHERE key = $1 AND expires_at > NOW()",
		key,
	).Scan(&response, &statusCode)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, 0, false, nil
	}
	if err != nil {
		return nil, 0, false, err
	}
	return response, statusCode, true, nil
}

// Store saves response for idempotency key with 24h TTL.
func (s *IdempotencyStore) Store(ctx context.Context, key string, response []byte, statusCode int) error {
	if s == nil || key == "" {
		return nil
	}
	_, err := s.pool.Exec(ctx,
		`INSERT INTO idempotency_keys (key, response, status_code, expires_at)
		 VALUES ($1, $2, $3, $4)
		 ON CONFLICT (key) DO UPDATE SET response = $2, status_code = $3`,
		key, response, statusCode, time.Now().Add(24*time.Hour),
	)
	return err
}

// Cleanup removes expired keys. Call periodically.
func (s *IdempotencyStore) Cleanup(ctx context.Context) error {
	if s == nil {
		return nil
	}
	_, err := s.pool.Exec(ctx, "DELETE FROM idempotency_keys WHERE expires_at < NOW()")
	return err
}

// NewIdempotencyInterceptor returns a connect-go interceptor that checks for
// duplicate requests using the Idempotency-Key header.
// If store is nil, the interceptor is a no-op pass-through.
func NewIdempotencyInterceptor(store *IdempotencyStore) connect.Interceptor {
	return connect.UnaryInterceptorFunc(func(next connect.UnaryFunc) connect.UnaryFunc {
		return func(ctx context.Context, req connect.AnyRequest) (connect.AnyResponse, error) {
			if store == nil {
				return next(ctx, req)
			}
			key := req.Header().Get("Idempotency-Key")
			if key == "" {
				return next(ctx, req)
			}
			// Check for cached response
			cached, statusCode, found, err := store.Check(ctx, key)
			if err == nil && found {
				_ = statusCode
				_ = cached
				// For connect-go, we can't easily reconstruct the response from bytes
				// So just log and proceed (full implementation would cache protobuf)
				slog.Debug("idempotency key found, but connect-go response reconstruction not supported", "key", key)
			}
			return next(ctx, req)
		}
	})
}
